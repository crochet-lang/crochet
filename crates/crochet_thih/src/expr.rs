use std::collections::HashSet;

use super::assump::*;
use super::context::*;
use super::id::*;
use super::pat::*;
use super::pred::*;
use super::prim::*;
use super::r#type::*;
use super::scheme::*;
use super::subst::*;

// data Expr = Var   Id
//           | Lit   Literal
//           -- named constant, i.e. the type of which is pre-determined
//           | Const Assump
//           | Ap    Expr Expr
//           | Let   BindGroup Expr

//           | Lam   Alt
//           | If    Expr Expr Expr
//           | Case  Expr [(Pat,Expr)]

pub enum Expr {
    Var(ID),
    Const(Assump),
    App(Box<Expr>, Vec<Expr>),
    Let(BindGroup, Box<Expr>),
    Lam(Alt),
    If(Box<Expr>, Box<Expr>, Box<Expr>),
    Case(Box<Expr>, Vec<(Pat, Box<Expr>)>),
}

// tiExpr                       :: Infer Expr Type
// tiExpr ce as (Var i)          = do sc         <- find i as
//                                    (ps :=> t) <- freshInst sc
//                                    return (ps, t)
pub fn ti_expr(ctx: &mut Context, r#as: &[Assump], e: &Expr) -> (Vec<Pred>, Type) {
    match e {
        Expr::Var(id) => {
            let a = r#as.iter().find(|a| &a.0 == id).unwrap();
            let sc = &a.1;
            let (ps, t) = ctx.fresh_inst(sc);
            (ps, t)
        }
        Expr::Const((_id, sc)) => {
            let (ps, t) = ctx.fresh_inst(sc);
            (ps, t)
        }
        Expr::App(r#fn, args) => {
            let mut all_preds: Vec<Pred> = vec![];

            let (mut fn_pred, fn_type) = ti_expr(ctx, r#as, r#fn);

            let mut arg_types: Vec<Type> = vec![];
            for arg in args {
                let (mut arg_preds, arg_type) = ti_expr(ctx, r#as, arg);
                all_preds.append(&mut arg_preds);
                arg_types.push(arg_type);
            }
            let ret_type = ctx.new_tvar();

            let call_type = Type {
                usage: Some(Usage::FnCall),
                variant: Variant::Lam(arg_types, Box::from(ret_type.clone())),
            };

            unify(ctx, &call_type, &fn_type);

            all_preds.append(&mut fn_pred);

            (all_preds, ret_type)
        }
        Expr::Let(bg, e) => {
            let (ps, as1) = ti_bind_group(ctx, r#as, bg);
            let (qs, t) = ti_expr(ctx, &plus_plus(r#as, &as1), e);
            (plus_plus(&ps, &qs), t)
        }
        Expr::Lam(alt) => ti_alt(ctx, r#as, alt),
        Expr::If(cond, e1, e2) => {
            let (ps, t) = ti_expr(ctx, r#as, cond);
            let t_bool = Type {
                usage: None,
                variant: Variant::Prim(Prim::Bool),
            };
            unify(ctx, &t, &t_bool);
            let (ps1, t1) = ti_expr(ctx, r#as, e1);
            let (ps2, t2) = ti_expr(ctx, r#as, e2);
            unify(ctx, &t1, &t2);
            (plus_plus(&plus_plus(&ps, &ps1), &ps2), t1)
        }
        Expr::Case(_, _) => todo!(),
    }
}

// -----------------------------------------------------------------------------

// type Alt = ([Pat], Expr)
type Alt = (Vec<Pat>, Box<Expr>);

// type Expl = (Id, Scheme, [Alt])
// NOTE: each implicitly typed binding has a single alternative in crochet
type Expl = (ID, Scheme, Vec<Alt>);

pub fn plus_plus<T: Clone>(v1: &[T], v2: &[T]) -> Vec<T> {
    v1.iter().chain(v2).cloned().collect()
}

// -----------------------------------------------------------------------------

// type Impl   = (Id, [Alt])
// NOTE: each implicitly typed binding has a single alternative in crochet
type Impl = (ID, Vec<Alt>);

// restricted   :: [Impl] -> Bool
// restricted bs = any simple bs
//  where simple (i,alts) = any (null . fst) alts
pub fn restricted(bs: &[Impl]) -> bool {
    bs.iter()
        .any(|(_i, alts)| alts.iter().any(|(fst, _snd)| fst.is_empty()))
}

// tiImpls         :: Infer [Impl] [Assump]
pub fn ti_impls(ctx: &mut Context, r#as: &[Assump], bs: &[Impl]) -> (Vec<Pred>, Vec<Assump>) {
    let ts: Vec<_> = bs.iter().map(|_| ctx.new_tvar()).collect();
    let r#is: Vec<String> = bs.iter().map(|(fst, _)| fst).cloned().collect();
    let scs: Vec<_> = ts.iter().map(to_scheme).collect();
    let as1 = plus_plus(&(r#is.iter().cloned().zip(scs).collect::<Vec<_>>()), r#as);
    let altss: Vec<_> = bs.iter().map(|(_, snd)| snd).collect();
    let pss: Vec<_> = altss
        .iter()
        .zip(&ts)
        .flat_map(|(alts, t)| ti_alts(ctx, &as1, alts, t))
        .collect();
    let s = ctx.get_subst();
    let ps1 = pss.apply(&s);
    let ts1 = ts.apply(&s);
    let fs = r#as.to_owned().apply(&s).tv();
    let vss: Vec<_> = ts1.iter().map(|t| t.tv()).collect();
    // gs      = foldr1 union vss \\ fs
    let gs = vss // TODO: make this vss \\ fs
        .iter()
        .cloned()
        .reduce(|accum, x| accum.union(&x).cloned().collect())
        .unwrap();

    let shared_vs = vss
        .iter()
        .cloned()
        .reduce(|accum, x| accum.intersection(&x).cloned().collect())
        .unwrap();
    let shared_vs: Vec<_> = shared_vs.iter().cloned().collect();

    let (ds, rs) = split(&fs, &shared_vs, &ps1);

    //     if restricted bs then
    if restricted(bs) {
        //     let gs'  = gs \\ tv rs
        let gs1: Vec<_> = gs.difference(&rs.tv()).cloned().collect();
        //         scs' = map (quantify gs' . ([]:=>)) ts'
        let scs1 = ts1.iter().map(|t| quantify(&gs1, &(vec![], t.to_owned())));
        //     in return (ds++rs, zipWith (:>:) is scs')
        (
            plus_plus(&ds, &rs),
            r#is.iter().cloned().zip(scs1).collect(),
        )
    } else {
        //     let scs' = map (quantify gs . (rs:=>)) ts'
        let scs1 = ts1.iter().map(|t| {
            quantify(
                &gs.iter().cloned().collect::<Vec<_>>(),
                &(rs.clone(), t.to_owned()),
            )
        });
        //     in return (ds, zipWith (:>:) is scs')
        (ds, r#is.iter().cloned().zip(scs1).collect())
    }
}

// tiAlt                :: Infer Alt Type
// tiAlt ce as (pats, e) = do (ps, as', ts) <- tiPats pats
//                            (qs,t)  <- tiExpr ce (as'++as) e
//                            return (ps++qs, foldr fn t ts)
pub fn ti_alt(ctx: &mut Context, r#as: &[Assump], (pats, e): &Alt) -> (Vec<Pred>, Type) {
    let (ps, as_1, ts) = ti_pats(ctx, pats);
    let (qs, t) = ti_expr(ctx, &plus_plus(&as_1, r#as), e);
    (
        // ps++qs
        plus_plus(&ps, &qs),
        // foldr fn t ts
        Type {
            usage: Some(Usage::Alt),
            variant: Variant::Lam(ts, Box::from(t)),
        },
    )
}

// tiAlts             :: ClassEnv -> [Assump] -> [Alt] -> Type -> TI [Pred]
// tiAlts ce as alts t = do psts <- mapM (tiAlt ce as) alts
//                          mapM (unify t) (map snd psts)
//                          return (concat (map fst psts))
pub fn ti_alts(ctx: &mut Context, r#as: &[Assump], alts: &[Alt], t: &Type) -> Vec<Pred> {
    let psts: Vec<_> = alts.iter().map(|alt| ti_alt(ctx, r#as, alt)).collect();
    for (_fst, snd) in psts.clone() {
        unify(ctx, t, &snd);
    }
    psts.iter().flat_map(|(fst, _snd)| fst).cloned().collect()
}

// tiExpl :: ClassEnv -> [Assump] -> Expl -> TI [Pred]
// tiExpl ce as (i, sc, alts)
pub fn ti_expl(ctx: &mut Context, r#as: &[Assump], (_i, sc, alts): &Expl) -> Vec<Pred> {
    let (qs, t) = ctx.fresh_inst(sc);
    let ps = ti_alts(ctx, r#as, alts, &t);
    let s = ctx.get_subst();
    let qs_1 = qs.apply(&s);
    let t_1 = t.apply(&s);
    // gs = tv t' \\ fs
    let fs = r#as.to_vec().apply(&s).tv();
    let gs: Vec<_> = t_1.apply(&s).tv().difference(&fs).cloned().collect();
    let sc_1 = quantify(&gs, &(qs_1, t_1));

    // ps' = filter (not . entail ce qs') (apply s ps)
    // NOTE: since `entail` always returns `false`, we can drop the `filter` call.
    let ps_1: Vec<_> = ps.apply(&s);
    let (ds, rs) = split(&fs, &gs, &ps_1);
    if sc != &sc_1 {
        panic!("signature too general");
    } else if !rs.is_empty() {
        panic!("context too weak");
    } else {
        ds
    }
}

// split :: Monad m => ClassEnv -> [Tyvar] -> [Tyvar] -> [Pred]
//                       -> m ([Pred], [Pred])
pub fn split(fs: &HashSet<ID>, _gs: &[ID], ps: &[Pred]) -> (Vec<Pred>, Vec<Pred>) {
    // split ce fs gs ps = do ps' <- reduce ce ps
    //                        let (ds, rs) = partition (all (`elem` fs) . tv) ps'
    //                        rs' <- defaultedPreds ce (fs++gs) rs
    //                        return (ds, rs \\ rs')
    // "deferred"
    let mut ds: Vec<Pred> = vec![];
    // "retained"
    let mut rs: Vec<Pred> = vec![];
    for p in ps {
        if p.tv().iter().all(|p_tv| fs.contains(p_tv)) {
            ds.push(p.to_owned());
        } else {
            rs.push(p.to_owned());
        }
    }
    // TODO: defaultedPreds
    // rs' <- defaultedPreds ce (fs++gs) rs
    // return (ds, rs \\ rs')
    (ds, rs)
}

// type Ambiguity       = (Tyvar, [Pred])
type Ambiguity = (ID, Vec<Pred>);

// ambiguities         :: ClassEnv -> [Tyvar] -> [Pred] -> [Ambiguity]
pub fn ambiguities(vs: &[ID], ps: &[Pred]) -> Vec<Ambiguity> {
    let vs: HashSet<_> = vs.iter().cloned().collect();
    ps.to_owned()
        .tv()
        .difference(&vs)
        .map(|v| {
            // (v, filter (elem v . tv) ps)
            (
                v.to_owned(),
                ps.iter().filter(|p| p.tv().contains(v)).cloned().collect(),
            )
        })
        .collect()
}

// -----------------------------------------------------------------------------

// type BindGroup  = ([Expl], [[Impl]])
type BindGroup = (Vec<Expl>, Vec<Vec<Impl>>);

pub fn ti_bind_group(
    ctx: &mut Context,
    r#as: &[Assump],
    (es, iss): &BindGroup,
) -> (Vec<Pred>, Vec<Assump>) {
    // do let as' = [ v:>:sc | (v,sc,alts) <- es ]
    let as1: Vec<Assump> = es
        .iter()
        .map(|(v, sc, _alts)| (v.to_owned(), sc.to_owned()))
        .collect();

    // (ps, as'') <- tiSeq tiImpls ce (as'++as) iss
    let (ps, as2) = ti_seq(ctx, &plus_plus(&as1, r#as), iss);

    // qss        <- mapM (tiExpl ce (as''++as'++as)) es
    let qss: Vec<_> = es
        .iter()
        .flat_map(|e| ti_expl(ctx, &plus_plus(&plus_plus(&as2, &as1), r#as), e))
        .collect();

    // return (ps++concat qss, as''++as')
    (plus_plus(&ps, &qss), plus_plus(&as2, &as1))
}

// tiSeq                  :: Infer bg [Assump] -> Infer [bg] [Assump]
pub fn ti_seq(ctx: &mut Context, r#as: &[Assump], bgs: &[Vec<Impl>]) -> (Vec<Pred>, Vec<Assump>) {
    let (all_preds, all_assumps): (Vec<Pred>, Vec<Assump>) = match bgs {
        [bs, bss @ ..] => {
            let (ps, as_1) = ti_impls(ctx, r#as, bs);
            let (qs, as_2) = ti_seq(ctx, &plus_plus(r#as, &as_1), bss);
            (plus_plus(&ps, &qs), plus_plus(&as_2, &as_1))
        }
        [] => (vec![], vec![]),
    };
    (all_preds, all_assumps)
}
