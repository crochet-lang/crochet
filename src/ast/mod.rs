pub mod literal;
pub mod syntax;
