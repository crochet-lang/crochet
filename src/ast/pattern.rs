use crate::ast::span::Span;
use crate::ast::ident::Ident;
use crate::ast::types::TypeAnn;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Pattern {
    Ident(BindingIdent),
    Rest(RestPat),

    // Array(ArrayPat),
    // Object(ObjectPat),
    // Assign(AssignPat),
}

impl Pattern {
    pub fn span(&self) -> Span {
        match self {
            Pattern::Ident(ident) => ident.span.to_owned(),
            Pattern::Rest(rest) => rest.span.to_owned(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BindingIdent {
    pub span: Span,
    pub id: Ident,
    pub type_ann: Option<TypeAnn>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RestPat {
    pub span: Span,
    pub arg: Box<Pattern>,
    pub type_ann: Option<TypeAnn>,
}
