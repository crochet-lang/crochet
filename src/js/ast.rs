use super::super::literal::Lit;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program {
    pub body: Vec<Statement>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Statement {
    Decl {
        pattern: Pattern,
        value: Expression,
    },
    Expression {
        expr: Expression,
    },
    Return {
        arg: Expression,   
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expression {
    Call {
        func: Box<Expression>,
        args: Vec<Expression>,
    },
    Function {
        params: Vec<Param>,
        body: Vec<Statement>,
        r#async: bool,
    },
    Ident {
        name: String,
    },
    Literal(Lit),
    Binary {
        op: BinaryOp,
        left: Box<Expression>,
        right: Box<Expression>,
    },
    Unary {
        op: UnaryOp,
        arg: Box<Expression>,
    },
    IfElse {
        cond: Box<Expression>,
        consequent: Vec<Statement>,
        alternate: Vec<Statement>,
    },
    Object {
        properties: Vec<Property>,
    },
    Await {
        expr: Box<Expression>,
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Pattern {
    Ident { name: String },
    // TODO: add more patterns later
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Param {
    Ident { name: String },
    Rest { name: String },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Property {
    pub name: String,
    pub value: Expression,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Exp,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum UnaryOp {
    Neg,
    Not,
}
