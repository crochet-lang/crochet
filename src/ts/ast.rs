use itertools::{join, Itertools};
use std::fmt;

use super::super::literal::Literal;
use super::super::types::Primitive;

pub struct TsQualifiedType {
    pub ty: TsType,
    pub type_params: Vec<i32>,
}

#[derive(Debug)]
pub enum TsType {
    Prim(Primitive),
    Var(String),
    Lit(Literal),
    Func {
        params: Vec<Param>,
        ret: Box<TsType>,
    },
    Union(Vec<TsType>),
    Obj(Vec<TsObjProp>),
    Alias {
        name: String,
        type_params: Vec<TsType>,
    },
}

impl fmt::Display for TsQualifiedType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let chars: Vec<_> = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
            .chars()
            .collect();

        if self.type_params.len() > 0 {
            let type_params = self
                .type_params
                .iter()
                .map(|id| {
                    let id = chars.get(id.to_owned() as usize).unwrap();
                    format!("{id}")
                })
                .join(", ");
            write!(f, "<{type_params}>{}", self.ty)
        } else {
            write!(f, "{}", self.ty)
        }
    }
}

impl fmt::Display for TsType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            TsType::Var(name) => write!(f, "{}", name),
            TsType::Prim(prim) => write!(f, "{}", prim),
            TsType::Lit(lit) => write!(f, "{}", lit),
            TsType::Func { params, ret } => {
                let params = params
                    .iter()
                    // TODO: use write! to format the params more directly instead of
                    // using the intermediary format!
                    .map(|Param { name, ty }| format!("{name}: {ty}").to_owned())
                    .join(", ");
                write!(f, "({params}) => {ret}")
            }
            TsType::Union(types) => write!(f, "{}", join(types, " | ")),
            TsType::Obj(props) => {
                let props = props
                    .iter()
                    .map(|TsObjProp { name, ty }| format!("{name}: {ty}"))
                    .join(", ");

                // TODO: output multi-line object types
                write!(f, "{{{props}}}")
            }
            TsType::Alias {name, type_params} => {
                let type_params = type_params.iter().join(", ");
                write!(f, "{name}<{type_params}>")
            }
        }
    }
}

#[derive(Debug)]
pub struct Param {
    pub name: String,
    pub ty: TsType,
}

#[derive(Debug)]
pub struct TsObjProp {
    pub name: String,
    pub ty: TsType,
}
