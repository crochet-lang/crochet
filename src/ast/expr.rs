use crate::ast::ident::Ident;
use crate::ast::jsx::JSXElement;
use crate::ast::literal::Lit;
use crate::ast::pattern::Pattern;
use crate::ast::span::Span;
use crate::ast::types::{TypeAnn, TypeParam};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program {
    pub body: Vec<Statement>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Statement {
    VarDecl {
        span: Span,
        pattern: Pattern,
        init: Option<Expr>,
        declare: bool,
    },
    TypeDecl {
        span: Span,
        declare: bool,
        id: Ident,
        type_ann: TypeAnn,    
        type_params: Option<Vec<TypeParam>>,    
    },
    Expr {
        span: Span,
        expr: Expr,
    }, // NOTE: does not include Expr::Let
}

// #[derive(Debug, Clone, PartialEq, Eq)]
// struct Block {
//     pub span: Span,
//     pub stmts: Vec<Expr>,
// }

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct App {
    pub span: Span,
    pub lam: Box<Expr>,
    pub args: Vec<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Fix {
    pub span: Span,
    pub expr: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct IfElse {
    pub span: Span,
    pub cond: Box<Expr>,
    pub consequent: Box<Expr>,
    pub alternate: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Lambda {
    pub span: Span,
    pub params: Vec<Pattern>,
    pub body: Box<Expr>,
    pub is_async: bool,
    pub return_type: Option<TypeAnn>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Let {
    pub span: Span,
    pub pattern: Option<Pattern>,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Op {
    pub span: Span,
    pub op: BinOp,
    pub left: Box<Expr>,
    pub right: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Obj {
    pub span: Span,
    pub props: Vec<Property>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Await {
    pub span: Span,
    pub expr: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Tuple {
    pub span: Span,
    pub elems: Vec<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Member {
    pub span: Span,
    pub obj: Box<Expr>,
    pub prop: MemberProp,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MemberProp {
    Ident(Ident),
    Computed(ComputedPropName),
}

impl MemberProp {
    pub fn name(&self) -> String {
        match self {
            MemberProp::Ident(Ident { name, .. }) => name.to_owned(),
            MemberProp::Computed(_) => todo!(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ComputedPropName {
    pub span: Span, // includes enclosing []
    pub expr: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expr {
    App(App),
    Fix(Fix),
    Ident(Ident),
    IfElse(IfElse),
    JSXElement(JSXElement),
    Lambda(Lambda),
    Let(Let),
    Lit(Lit),
    Op(Op),
    Obj(Obj),
    Await(Await),
    Tuple(Tuple),
    Member(Member),
}

impl Expr {
    pub fn span(&self) -> Span {
        match &self {
            Expr::App(app) => app.span.to_owned(),
            Expr::Fix(fix) => fix.span.to_owned(),
            Expr::Ident(ident) => ident.span.to_owned(),
            Expr::IfElse(if_else) => if_else.span.to_owned(),
            Expr::JSXElement(elem) => elem.span.to_owned(),
            Expr::Lambda(lam) => lam.span.to_owned(),
            Expr::Let(r#let) => r#let.span.to_owned(),
            Expr::Lit(lit) => lit.span(),
            Expr::Op(op) => op.span.to_owned(),
            Expr::Obj(obj) => obj.span.to_owned(),
            Expr::Await(r#await) => r#await.span.to_owned(),
            Expr::Tuple(tuple) => tuple.span.to_owned(),
            Expr::Member(member) => member.span.to_owned(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    EqEq,
    NotEq,
    Gt,
    GtEq,
    Lt,
    LtEq,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Property {
    pub span: Span,
    pub name: String,
    pub value: Expr,
}
