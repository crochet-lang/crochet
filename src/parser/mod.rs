pub mod expr;
pub mod jsx;
pub mod pattern;
pub mod types;

use expr::expr_parser;
use pattern::pattern_parser;

use chumsky::prelude::*;
use chumsky::primitive::*;
use chumsky::text::Padded;

use crate::ast::*;

pub fn just_with_padding(inputs: &str) -> Padded<Just<char, &str, Simple<char>>> {
    just(inputs).padded()
}

pub fn parser() -> impl Parser<char, Program, Error = Simple<char>> {
    // We use `just` instead of `just_with_padding` here to ensure that
    // the span doesn't include leading whitespace.
    // TODO: add `declare` syntax
    let decl = just("let")
        .ignore_then(just_with_padding("rec").or_not())
        .then(pattern_parser())
        .then_ignore(just_with_padding("="))
        .then(expr_parser())
        .map_with_span(|((rec, pattern), value), span: Span| -> Statement {
            match rec {
                Some(_) => {
                    Statement::Decl {
                        span,
                        pattern: pattern.clone(),
                        // `let fib = fix((fib) => (n) => ...)`
                        // TODO: Fix always wraps a lambda
                        value: Expr::Fix(Fix {
                            span: value.span(),
                            expr: Box::from(Expr::Lambda(Lambda {
                                span: value.span(),
                                args: vec![pattern.clone()],
                                body: Box::from(value),
                                is_async: false,
                            })),
                        }),
                    }
                }
                None => Statement::Decl {
                    span,
                    pattern,
                    value,
                },
            }
        });

    let program = choice((
        decl,
        expr_parser().map_with_span(|expr, span: Span| Statement::Expr { expr, span }),
    ))
    .padded()
    .repeated()
    .map(|body| Program { body });

    program.then_ignore(end())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse(input: &str) -> Program {
        parser().parse(input).unwrap()
    }

    #[test]
    fn literals() {
        insta::assert_debug_snapshot!(parse("10"));
        insta::assert_debug_snapshot!(parse("1.23"));
        insta::assert_debug_snapshot!(parse("\"hello\""));
    }

    #[test]
    fn operations() {
        insta::assert_debug_snapshot!(parse("1 + 2 - 3"));
        insta::assert_debug_snapshot!(parse("x * y / z"));
        insta::assert_debug_snapshot!(parse("(a + b) * c"));
        insta::assert_debug_snapshot!(parse("a == b"));
        insta::assert_debug_snapshot!(parse("a != b"));
        insta::assert_debug_snapshot!(parse("a > b"));
        insta::assert_debug_snapshot!(parse("a >= b"));
        insta::assert_debug_snapshot!(parse("a < b"));
        insta::assert_debug_snapshot!(parse("a <= b"));
        insta::assert_debug_snapshot!(parse("let cond = a != b"));
    }

    #[test]
    fn function_definition() {
        insta::assert_debug_snapshot!(parse("(a, b) => c"));
        insta::assert_debug_snapshot!(parse("() => 10"));
        insta::assert_debug_snapshot!(parse("(a) => \"hello\""));
    }

    #[test]
    fn async_await() {
        insta::assert_debug_snapshot!(parse("async () => 10"));
        insta::assert_debug_snapshot!(parse("let foo = async () => { await 10 }"));
        insta::assert_debug_snapshot!(parse("let foo = async () => await a + await b"));
        insta::assert_debug_snapshot!(parse("let foo = async () => await bar()"));
    }

    #[test]
    fn function_application() {
        insta::assert_debug_snapshot!(parse("foo()"));
        insta::assert_debug_snapshot!(parse("foo(a, b)"));
        insta::assert_debug_snapshot!(parse("foo(10, \"hello\")"));
        insta::assert_debug_snapshot!(parse("f(x)(g(x))"));
    }

    #[test]
    fn declarations() {
        insta::assert_debug_snapshot!(parse("let x = 5"));
        insta::assert_debug_snapshot!(parse("let x = (a, b) => a + b"));
        insta::assert_debug_snapshot!(parse("let foo = let x = 5 in x"));
        insta::assert_debug_snapshot!(parse("let rec f = () => f()")); // recursive
    }

    #[test]
    fn top_level_expressions() {
        insta::assert_debug_snapshot!(parse("a + b"));
        insta::assert_debug_snapshot!(parse("123\n\"hello\""));
    }

    #[test]
    fn if_else() {
        insta::assert_debug_snapshot!(parse("if (true) { 5 } else { 10 }"));
    }

    #[test]
    fn objects() {
        insta::assert_debug_snapshot!(parse("{x: 5, y: 10}"));
    }

    #[test]
    fn jsx() {
        insta::assert_debug_snapshot!(parse("<Foo>Hello</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo>{bar}</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo>Hello {world}!</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo>{<Bar>{baz}</Bar>}</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo></Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo bar={baz} />"));
        insta::assert_debug_snapshot!(parse("<Foo msg=\"hello\" bar={baz}></Foo>"));
    }

    #[test]
    fn type_annotations() {
        insta::assert_debug_snapshot!(parse("let x: number = 5"));
        insta::assert_debug_snapshot!(parse("let msg: string = \"hello\""));
        insta::assert_debug_snapshot!(parse("let add = (a: number, b: number) => a + b"));
        insta::assert_debug_snapshot!(parse("let p: Point = {x: 5, y: 10}"));
        insta::assert_debug_snapshot!(parse("let FOO: \"foo\" = \"foo\""));
    }
}
