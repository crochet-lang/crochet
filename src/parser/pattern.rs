use chumsky::prelude::*;

use crate::ast::*;
use crate::parser::types::*;
use crate::parser::util::just_with_padding;

// NOTE: Destructuring assignments admits different patterns from destructuring
// function params.  We'll need to have different parsers for those.
pub fn pattern_parser() -> BoxedParser<'static, char, Pattern, Simple<char>> {
    let type_ann = type_parser();
    let mut top_level = true;

    let parser = recursive(|pat| {
        let ident_pat = text::ident()
            .map_with_span(|name, span| Ident { name, span })
            .then(just_with_padding(":").ignore_then(type_ann).or_not())
            .map_with_span(|(id, type_ann), span: Span| {
                Pattern::Ident(BindingIdent { span, id, type_ann })
            })
            .padded();

        // NOTE: rest patterns must appear instead of other patterns
        // It's an error for them appear at the top-level, e.g.
        // let ...x = [1,2,3];
        // Also, type annotation can only appear at the top-level as well.
        let rest_pat = just("...")
            .ignore_then(pat.clone())
            .map_with_span(|arg, span| RestPat {
                span,
                arg: Box::from(arg),
                type_ann: None,
            });

        let array_pat = pat.clone()
            .separated_by(just_with_padding(","))
            .delimited_by(just_with_padding("["), just_with_padding("]"))
            .map_with_span(|elems, span| {
                Pattern::Array(ArrayPat {
                    span,
                    // The reason why each elem is wrapped in Some() is that
                    // it's possible to have a pattern with skipped elements
                    // althought the parser doesn't support this yet.
                    elems: elems.iter().cloned().map(Some).collect(),
                    optional: false,
                    type_ann: None,
                })
            });

        let key_value_pat_prop = text::ident()
            .map_with_span(|name, span| Ident { span, name })
            .then_ignore(just_with_padding(":"))
            .then(pat.clone())
            .map(|(key, value)| {
                ObjectPatProp::KeyValue(KeyValuePatProp {
                    key,
                    value: Box::from(value),
                })
            });

        // TODO: support default values
        let assign_pat_prop = text::ident()
            .map_with_span(|name, span| Ident { span, name })
            .map_with_span(|key, span| {
                ObjectPatProp::Assign(AssignPatProp {
                    span,
                    key,
                    value: None,
                })
            });

        // NOTE: There can only be a single rest element and it must be last
        let obj_pat_prop = choice((
            rest_pat.clone().map(ObjectPatProp::Rest),
            key_value_pat_prop,
            assign_pat_prop,
        ))
        .separated_by(just_with_padding(","))
        .delimited_by(just_with_padding("{"), just_with_padding("}"))
        .map_with_span(|props, span| {
            Pattern::Object(ObjectPat {
                span,
                props,
                optional: false,
                type_ann: None,
            })
        });

        top_level = false;

        choice((ident_pat, array_pat, obj_pat_prop, rest_pat.map(Pattern::Rest)))
    });

    parser.boxed()
}
