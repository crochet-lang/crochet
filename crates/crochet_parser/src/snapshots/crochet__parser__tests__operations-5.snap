---
source: src/parser/mod.rs
expression: "parse(\"a != b\")"
---
Program {
    body: [
        Expr {
            span: 0..6,
            expr: Op(
                Op {
                    span: 0..6,
                    op: NotEq,
                    left: Ident(
                        Ident {
                            span: 0..1,
                            name: "a",
                        },
                    ),
                    right: Ident(
                        Ident {
                            span: 5..6,
                            name: "b",
                        },
                    ),
                },
            ),
        },
    ],
}
